source/common/config/delta_subscription_state.cc

```diff
@@ -14,52 +14,128 @@ namespace Config {
           },
           dispatcher, dispatcher.timeSource()),
-      type_url_(std::move(type_url)), wildcard_(wildcard), watch_map_(watch_map),
-      local_info_(local_info), dispatcher_(dispatcher) {}
+      type_url_(std::move(type_url)), watch_map_(watch_map), local_info_(local_info),
+      dispatcher_(dispatcher) {}
 
 void DeltaSubscriptionState::updateSubscriptionInterest(
     const absl::flat_hash_set<std::string>& cur_added,
     const absl::flat_hash_set<std::string>& cur_removed) {
   for (const auto& a : cur_added) {
-    setResourceWaitingForServer(a);
+    if (in_initial_legacy_wildcard_ && a != Wildcard) {
+      in_initial_legacy_wildcard_ = false;
+    }
+    // If the requested resource existed as a wildcard resource,
+    // transition it to requested. Otherwise mark it as a resource
+    // waiting for the server to receive the version.
+    if (auto it = wildcard_resource_state_.find(a); it != wildcard_resource_state_.end()) {
+      requested_resource_state_.insert_or_assign(a, ResourceState::withVersion(it->second));
+      wildcard_resource_state_.erase(it);
+    } else if (it = ambiguous_resource_state_.find(a); it != ambiguous_resource_state_.end()) {
+      requested_resource_state_.insert_or_assign(a, ResourceState::withVersion(it->second));
+      ambiguous_resource_state_.erase(it);
+    } else {
+      requested_resource_state_.insert_or_assign(a, ResourceState::waitingForServer());
+    }
+    ASSERT(requested_resource_state_.contains(a));
+    ASSERT(!wildcard_resource_state_.contains(a));
+    ASSERT(!ambiguous_resource_state_.contains(a));
     // If interest in a resource is removed-then-added (all before a discovery request
     // can be sent), we must treat it as a "new" addition: our user may have forgotten its
     // copy of the resource after instructing us to remove it, and need to be reminded of it.
     names_removed_.erase(a);
     names_added_.insert(a);
   }
   for (const auto& r : cur_removed) {
-    removeResourceState(r);
+    auto actually_erased = false;
+    // The resource we have lost the interest in could also come from our wildcard subscription. We
+    // just don't know it at this point. Instead of removing it outright, mark the resource as not
+    // interesting to us any more and the server will send us an update. If we don't have a wildcard
+    // subscription then there is no ambiguity and just drop the resource.
+    if (requested_resource_state_.contains(Wildcard)) {
+      if (auto it = requested_resource_state_.find(r); it != requested_resource_state_.end()) {
+        // Wildcard resources always have a version. If our requested resource has no version, it
+        // won't be a wildcard resource then. If r is Wildcard itself, then it never has a version
+        // attached to it, so it will not be moved to ambiguous category.
+        if (!it->second.isWaitingForServer()) {
+          ambiguous_resource_state_.insert({it->first, it->second.version()});
```

Comment:
Just a short note before I reply to the rest of the message - this piece of code is executed when client changes its subscription interest, not when server sends the client a message about resource changes. This seems to be your source of confusion here.

> it's a little bit unclear to me if how to resolve the ambiguity here:

For the record, ambiguity is on the client side, and happens when client is subscribed to `A` and `*` at the same, and then decides to unsubscribe from `A`. At this point the client doesn't know if `A` was also a part of the wildcard resource subscription. Thus, `A` is ambiguous, because it might be coming from wildcard resource set or should be dropped from the cache.

> e.g., client requested EDS cluster "cluster_A" and also is a "wildcard" client,
> Server sends deletion of "cluster_A", cluster_A destructs and is inserted into "ambiguous_resource_state_" and later removed from "ambiguous_resource_state_".

The insertion to `ambiguous_resource_state_` happens only due to the subscription interest change, not as a reaction to server responses. When client is subscribed to `A` and `*`, and then unsubscribes from `A`, it inserts `A` into `ambiguous_resource_state_`. On the other hand, in your scenario, when server sends deletion of `cluster_A`, the client marks `cluster_A` as "waiting for server" instead of deleting it (because the client is still interested in `cluster_A` despite it being deleted). So in this case, `cluster_A` remains in `requested_resource_state_`. This is coded in the `handleGoodResponse` function.

> And it's also removed from "requested_resource_state_"?

A resource can be only in one of `requested_resource_state_`, `ambiguous_resource_state_` and `wildcard_resource_state_` at a time. So yes, if it's being put it `ambiguous_resource_state_` then will be removed from `requested_resource_state_`.

> In that case, server is deleting explicitly requested resources(even if it has cluster_A later), is this desired?
> 
> I'd admit it's uncommon that EDS been "wildcard" tho, so probably fine.

As hopefully clarified above, it isn't what happens here.

Reply:
is this wildcard only for previously "wildcard" XDSes? if so this might not be a concern, but we need to make it clear in the doc.

the issue I am trying to understand better is that:  for an EDS cluster, when a CDS update removes it, it will "unsubscribe" and update interest in EDS. where this code will be called, see how EdsClusterImpl::subscription_ destructs leading to unsubscribing the eds resource.

Reply:
> is this wildcard only for previously "wildcard" XDSes? if so this might not be a concern, but we need to make it clear in the doc.

XDSes that previously didn't have a wildcard subscription before this change still won't automatically get it now. They need to opt-in by expressing interest in `*` resource.

> the issue I am trying to understand better is that: for an EDS cluster, when a CDS update removes it, it will "unsubscribe" and update interest in EDS. where this code will be called, see how EdsClusterImpl::subscription_ destructs leading to unsubscribing the eds resource.

So EdsCluster has a subscription over which it expresses interest on just a single resource (see `EdsClusterImpl::startPreInit`). This resource will end up in the `requested_resource_state_`. This PR doesn't change anything that happens between the destruction of `EdsClusterImpl::subscription_` and invoking of `updateSubscriptionInterest`. So, if in your case EDS also has a wildcard subscription, then the resource will be moved to `ambiguous_resource_state_`. At some point, the server will let us know if the resource should be dropped from the cache or moved to `wildcard_resource_state_`.

But my understanding is that currently EDS does not use wildcard subscriptions, so when EdsCluster is destroyed, the resource is going to be removed from cache.

Reply:
yeah, my question is that if user somehow add * to EDS subscription and then removed it. Then the scenario described above will actually remove the "explicitly " requested EDS resources while it should not. 

sorry if I am asking some confusing questions.

could this sequence be possible:

1. EDS add A, * 
2. CDS from Server removes A, cause an EDS updateSubscriptionInterest({}, {A}), A is moved to ambiguous_resource_state_, added to names_removed_. 
3. EDS request sent to Server, with * in it
4. Client also removes *, clears ambiguous_resource_state_
5. Server sends back updated A", but A is not in the ambiguous_resource_state_, and eds is not wildcard anymore,
6. A is forgot in the stream. 

The above sequence is based on the assumption that * could be added to a EDS stream.

Reply:
It is not legal to use a wildcard subscription with EDS; that behavior is defined only for LDS and CDS:

https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#how-the-client-specifies-what-resources-to-return

Reply:
> yeah, my question is that if user somehow add * to EDS subscription and then removed it. Then the scenario described above will actually remove the "explicitly " requested EDS resources while it should not.
> 
> sorry if I am asking some confusing questions.
> 
> could this sequence be possible:
> 
>     1. EDS add A, *
> 
>     2. CDS from Server removes A, cause an EDS updateSubscriptionInterest({}, {A}), A is moved to ambiguous_resource_state_, added to names_removed_.
> 
>     3. EDS request sent to Server, with * in it
> 
>     4. Client also removes *, clears ambiguous_resource_state_
> 
>     5. Server sends back updated A", but A is not in the ambiguous_resource_state_, and eds is not wildcard anymore,
> 
>     6. A is forgot in the stream.
> 
> 
> The above sequence is based on the assumption that * could be added to a EDS stream.

Putting aside the legality of wildcard EDS, I'm not really that well versed between the interactions between EDS and CDS. Do EDS and CDS work on the same type of resources? Because it's confusing that both CDS and EDS refer to A as if it was the same resource. If EDS calls `updateSubscriptionInterest({}, {A})` then the intent of such a call is to "forget" about `A`.

Reply:
Each xDS API (LDS, RDS, CDS, EDS) uses a different resource type.  A resource is uniquely identified by both type and name, so it's fine to have both a CDS resource and an EDS resource of the same name.

Reply:
Right, so `A` from CDS may be in a different state (like waiting for server) than `A` from EDS (dropped from the cache).